### PR TITLE
nnjai support

### LINKED ISSUE #128: Add support for NNJA-AI Data
BrightBand has released the NNJA-AI dataset now here: https://www.brightband.com/data that is cloud-optimized data of observations from NOAA's data assimilation process. 

## Context
This could be a really nice way of training on pure observations or just including observations in a consistent format.

## Possible Implementation
They have a package that supports querying and getting the data. This would be wrapping that in a dataset to be able to load it on the fly, or generate examples on disk.

### PR DESCRIPTION
**Pull request by yuvraajnarula**
# Pull Request for issue #128 
@jacobbieker  

## Description  
This code provides a data pipeline for handling and processing satellite data using the NNJAI library. It focuses on creating a PyTorch Dataset class (`AMSUDataset`) for efficiently loading and accessing Advanced Microwave Sounding Unit (AMSU) data. Below is a detailed breakdown of its functionality:  

1. **Authentication Check**: Verifies access to NNJAI's `DataCatalog` using `_check_authentication()`.  
2. **Dynamic Dataset Initialization**: Loads and filters AMSU data based on specified primary descriptors and additional variables.  
3. **PyTorch Integration**: Implements `Dataset` and `DataLoader` for seamless compatibility with PyTorch-based pipelines.  
4. **Flexible Configuration**: Users can customize dataset parameters, timestamp filters, and metadata selection.  
5. **Fallback for Missing Dependencies**: Prompts users to install the NNJAI library if not already available.  

This implementation enhances satellite data processing workflows by providing an efficient and modular solution.

Fixes #128  

## How Has This Been Tested?  

_Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration._  

1. **Unit Tests**:  
   - Tested dataset loading with valid and invalid dataset names.  
   - Verified filtering based on `time`, `primary_descriptors`, and `additional_variables`.  
   - Checked the output format of `__getitem__` to ensure data integrity.  

2. **Integration Tests**:  
   - Used `DataLoader` to iterate over the dataset and verified the randomness of shuffled outputs.  

- [ ] Yes  

_If your changes affect data processing, have you plotted any changes? i.e. have you done a quick sanity check?_  

- [ ] Yes  

## Checklist:  

- [ ] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)  
- [ ] I have performed a self-review of my own code  
- [ ] I have made corresponding changes to the documentation  
- [ ] I have added tests that prove my fix is effective or that my feature works  
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by jacobbieker — CHANGES_REQUESTED** [review]
Thank you for this! This is a good start, I think there needs to be a few different changes before this can be merged, but a good start!

**Comment by yuvraajnarula** [comment]
Here is the output for the suggestions you requested.
```
(py310env) PS D:\code\os_contri\graph_weather> C:\Users\DELL\miniconda3\envs\py310env\python.exe graph_weather/data/nnjai_wrapp.py
Batch size: 4
Timestamps shape: torch.Size([4])
Latitudes shape: torch.Size([4])
Longitudes shape: torch.Size([4])
Metadata shape: torch.Size([4, 1])

All items in batch:

Item 0:
Time: 1609524352.0
Latitude: -31.94420051574707
Longitude: 60.85639953613281
Metadata: tensor([180.2300])

Item 1:
Time: 1609531264.0
Item 1:
Time: 1609531264.0
Latitude: -68.8927993774414
Item 1:
Time: 1609531264.0
Latitude: -68.8927993774414
Longitude: 57.41469955444336
Metadata: tensor([209.7500])
Item 1:
Time: 1609531264.0
Latitude: -68.8927993774414
Longitude: 57.41469955444336
Metadata: tensor([209.7500])

Item 1:
Time: 1609531264.0
Latitude: -68.8927993774414
Longitude: 57.41469955444336
Metadata: tensor([209.7500])

Item 2:
Latitude: -68.8927993774414
Longitude: 57.41469955444336
Metadata: tensor([209.7500])

Item 2:
Time: 1609520640.0
Latitude: -79.92410278320312
Longitude: 97.82360076904297
Metadata: tensor([164.9300])
Latitude: -68.8927993774414
Longitude: 57.41469955444336
Metadata: tensor([209.7500])

Item 2:
Time: 1609520640.0
Latitude: -79.92410278320312
Longitude: 97.82360076904297
Metadata: tensor([164.9300])

Item 3:
Time: 1609537536.0
Latitude: -14.215499877929688
Longitude: 179.46139526367188
Metadata: tensor([223.6900])
(py310env) PS D:\code\os_contri\graph_weather>
Latitude: -68.8927993774414
Longitude: 57.41469955444336
Metadata: tensor([209.7500])

Item 2:
Time: 1609520640.0
Latitude: -79.92410278320312
Longitude: 97.82360076904297
Metadata: tensor([164.9300])


Item 2:
Time: 1609520640.0
Latitude: -79.92410278320312
Longitude: 97.82360076904297
Metadata: tensor([164.9300])

Item 3:
Time: 1609537536.0
Latitude: -14.215499877929688
Longitude: 179.46139526367188
Metadata: tensor([223.6900])
```

**Comment by jacobbieker** [comment]
> Here is the output for the suggestions you requested.
> 
> ```
> (py310env) PS D:\code\os_contri\graph_weather> C:\Users\DELL\miniconda3\envs\py310env\python.exe graph_weather/data/nnjai_wrapp.py
> Batch size: 4
> Timestamps shape: torch.Size([4])
> Latitudes shape: torch.Size([4])
> Longitudes shape: torch.Size([4])
> Metadata shape: torch.Size([4, 1])
> 
> All items in batch:
> 
> Item 0:
> Time: 1609524352.0
> Latitude: -31.94420051574707
> Longitude: 60.85639953613281
> Metadata: tensor([180.2300])
> 
> Item 1:
> Time: 1609531264.0
> Item 1:
> Time: 1609531264.0
> Latitude: -68.8927993774414
> Item 1:
> Time: 1609531264.0
> Latitude: -68.8927993774414
> Longitude: 57.41469955444336
> Metadata: tensor([209.7500])
> Item 1:
> Time: 1609531264.0
> Latitude: -68.8927993774414
> Longitude: 57.41469955444336
> Metadata: tensor([209.7500])
> 
> Item 1:
> Time: 1609531264.0
> Latitude: -68.8927993774414
> Longitude: 57.41469955444336
> Metadata: tensor([209.7500])
> 
> Item 2:
> Latitude: -68.8927993774414
> Longitude: 57.41469955444336
> Metadata: tensor([209.7500])
> 
> Item 2:
> Time: 1609520640.0
> Latitude: -79.92410278320312
> Longitude: 97.82360076904297
> Metadata: tensor([164.9300])
> Latitude: -68.8927993774414
> Longitude: 57.41469955444336
> Metadata: tensor([209.7500])
> 
> Item 2:
> Time: 1609520640.0
> Latitude: -79.92410278320312
> Longitude: 97.82360076904297
> Metadata: tensor([164.9300])
> 
> Item 3:
> Time: 1609537536.0
> Latitude: -14.215499877929688
> Longitude: 179.46139526367188
> Metadata: tensor([223.6900])
> (py310env) PS D:\code\os_contri\graph_weather>
> Latitude: -68.8927993774414
> Longitude: 57.41469955444336
> Metadata: tensor([209.7500])
> 
> Item 2:
> Time: 1609520640.0
> Latitude: -79.92410278320312
> Longitude: 97.82360076904297
> Metadata: tensor([164.9300])
> 
> 
> Item 2:
> Time: 1609520640.0
> Latitude: -79.92410278320312
> Longitude: 97.82360076904297
> Metadata: tensor([164.9300])
> 
> Item 3:
> Time: 1609537536.0
> Latitude: -14.215499877929688
> Longitude: 179.46139526367188
> Metadata: tensor([223.6900])
> ```

Hi, could you push the changes to this branch? I don't see any differences in the code?

**Comment by yuvraajnarula** [comment]
@jacobbieker any updates?

**Review by jacobbieker — CHANGES_REQUESTED** [review]
Its getting there! The only thing I would go with changing is to add a unit test in the `tests/` directory that checks the output shapes from the dataset to ensure it is what it expects. You basically have the test here, under the __main__ bit, just change it to check for the output shapes and put that part in a unit test, then it can be merged!

**Review by jacobbieker — APPROVED** [review]
This looks good to me! Thanks for making the changes.

**Comment by yuvraajnarula** [comment]
Thank you for guiding me. If there's anything more I could do for this issue then do let me know.
